### PR TITLE
test: add the unit test of module_base/complexarray.h, and fix the bu…

### LINKED ISSUE #616: Incorrect using of "inline" function in module_base/complexarray.h
In module_base/complexarray.h, it has declared an "inline" function "inline void operator=(std::complex <double> c);", but the definition is in "complexarray.cpp", this will cause error when this function is called in other codes.

I suggest modify this function to be a normal function by removing "inline".

### PR DESCRIPTION
**Pull request by pxlxingliang**
…g of function "inline void operator=(std::complex <double> c)"

range: source/module_base

### CODE REVIEW

File: source/module_base/test/complexarray_test.cpp [hunk 1]
```diff
@@ -0,0 +1,207 @@
+#include"../complexarray.h"
+#include"gtest/gtest.h"
+
+/************************************************
+*  unit test of class ComplexArray
+***********************************************/
+
+/**
+* function "inline void ComplexArray::operator=(const std::complex < double> c)"
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp]: Please remove this if you have fixed these bug.

File: source/module_base/test/complexarray_test.cpp [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+*
+* the definition of this function should be in .h file, 
+* or do not define as an "inline" function, but a normal function.
+*/
+
+
+/**
+* - Tested Function:
+*   - operator "=":
+*       - assign a complex to all elements of a ComplexArray
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp:11]: The explanations here seems to be duplicated. I think the test name should satisfy.

File: source/module_base/test/complexarray_test.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+*       - one ComplexArray is multiplied by a complex
+*       - one ComplexArray is multiplied by another ComplexArray that has same dimension
+*       - throw error when one ComplexArray is miltiplied by another ComplexArray
+*         that has different dimension.
+*
+*   - operator "*=":
+*       similar as "*"
+*
+*   - oprator "()":
+*       - assess the element
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp]: typo here: should be 'access'

File: source/module_base/test/complexarray_test.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+            d2 = com1;
+            a2_plus_b2 = com1 + com2;
+            a2_minus_b2 = com1 - com2;
+            a2_mult_b2 = com1 * com2;
+            a2_mult_3 = com1 * 3.0;
+        }
+
+};
+
+bool Compare(ModuleBase::ComplexArray &ca1, ModuleBase::ComplexArray &ca2)
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp]: Consider adding this function to `ComplexArray` class as `operator ==`?

File: source/module_base/test/complexarray_test.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    return true;
+}
+
+
+TEST_F(ComplexArray_test,operator_equal)
+{   
+    c2 = a2;
+    b2 = a2;
+    EXPECT_TRUE(Compare(c2,a2));
+    EXPECT_TRUE(Compare(b2,a2));
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp]: What is the purpose to test twice?

**Review by caic99 — CHANGES_REQUESTED** [review]

**Comment by caic99** [comment]
Please split your test and bug-fixing into two separated PRs.

File: source/module_base/test/complexarray_test.cpp (hunk 2 above)
Comment (pxlxingliang) [reply source/module_base/test/complexarray_test.cpp:11]: > The explanations here seems to be duplicated. I think the test name should satisfy.

the test name can reflect a general function, but may not reflect the detail tested functions, I still think this commands should be  saved.

File: source/module_base/test/complexarray_test.cpp (hunk 4 above)
Comment (pxlxingliang) [reply source/module_base/test/complexarray_test.cpp]: > good idea, I can try to add this function

File: source/module_base/test/complexarray_test.cpp (hunk 5 above)
Comment (pxlxingliang) [reply source/module_base/test/complexarray_test.cpp]: b2 is assigned by a complex, while c2 is not assigned. These should be two different situations.

File: source/module_base/test/complexarray_test.cpp [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+            a2_mult_b2 = com1 * com2;
+            a2_mult_3 = com1 * 3.0;
+        }
+
+};
+
+TEST_F(ComplexArray_test,operator_equal)
+{   
+    c2 = a2; //c2 is not assigned
+    b2 = a2; //b2 is arrigned by a complex
```
Comment (caic99) [inline source/module_base/test/complexarray_test.cpp]: Typo here?

**Review by caic99 — CHANGES_REQUESTED** [review]

File: source/module_base/test/complexarray_test.cpp (hunk 5 above)
Comment (caic99) [reply source/module_base/test/complexarray_test.cpp]: > b2 is assigned by a complex, while c2 is not assigned. These should be two different situations.

Hi @pxlxingliang,
I don't think they are different situations, both cases are calling the same instructions. 
https://github.com/deepmodeling/abacus-develop/blob/9aa165087b1049eb668165a9c6e354336374e903/source/module_base/complexmatrix.cpp#L226-L231
Am I right? Or I've ignored something? Whatever, adding a comment explaining it is a good practice.

However, you did not test this function:
https://github.com/deepmodeling/abacus-develop/blob/9aa165087b1049eb668165a9c6e354336374e903/source/module_base/complexmatrix.cpp#L234-L242

Comment (pxlxingliang) [reply source/module_base/test/complexarray_test.cpp]: this function is actually not used in ABACUS, i think the test is unnecessary

**Review by caic99 — APPROVED** [review]

**Pull request merged**